snuba/clickhouse/query.py

```diff
@@ -38,7 +33,7 @@ def __init__(self,
         selected_cols = [column_expr(dataset, util.tuplify(colname), query, parsing_context) for colname in column_names]
         select_clause = u'SELECT {}'.format(', '.join(group_exprs + aggregate_exprs + selected_cols))
 
-        from_clause = u'FROM {}'.format(source)
+        from_clause = u'FROM {}'.format(query.get_from_clause().format())
```

Comment:
This is the key: Clickhouse query does not depend on dataset directly to build the FROM clause, which means query processors can tweak the FROM clause as they like.
The same approach will exists for columns thus removing the other dataset dependencies from this class.